reasoner/planner/AnswerCountEstimator.java

```diff
@@ -620,17 +629,31 @@ private LocalModel.RelationModel modelForRelation(RelationConstraint relationCon
         }
 
         private LocalModel.HasModel modelForHas(HasConstraint hasConstraint, @Nullable Concludable correspondingConcludable) {
-            double hasEdgeEstimate = countPersistedHasEdges(hasConstraint.owner().inferredTypes(), hasConstraint.attribute().inferredTypes());
+            double persistedHasEdgeEstimate = countPersistedHasEdges(hasConstraint.owner().inferredTypes(), hasConstraint.attribute().inferredTypes());
             double ownerEstimate = countPersistedThingsMatchingType(hasConstraint.owner());
             double attributeEstimate = countPersistedThingsMatchingType(hasConstraint.attribute());
+
+            double hasEdgeEstimate;
             if (correspondingConcludable != null) {
-                hasEdgeEstimate += estimateInferredAnswerCount(correspondingConcludable, set(hasConstraint.owner(), hasConstraint.attribute()));
                 attributeEstimate += attributesCreatedByExplicitHas(correspondingConcludable);
-                if (attributeEstimate != 0 && ownerEstimate < hasEdgeEstimate / attributeEstimate) {
+
+                double inferredHasEdgeCount = estimateInferredAnswerCount(correspondingConcludable, set(hasConstraint.owner(), hasConstraint.attribute()));
+
+                if (attributeEstimate != 0 && ownerEstimate < inferredHasEdgeCount / attributeEstimate) {
```

Comment:
I see this existed before this change, but it always scares me to compare `double`s with `==0`.

Reply:
Makes sense. It used to be a long but I didn't update these.
I can do a sweep now. Is there an easier way to do it than `Math.abs(x-y) < EPS` ?

Reply:
Only if you're confident about the sign.

Reply:
The sign of what?

Reply:
Of `x-y`, such that you don't have to take the `abs()`. For example, in these cases, `attributeEstimate > 0.0` or `attributeEstimate > -epsilon` might suffice.